cluster-autoscaler/loop/trigger.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package loop
+
+import (
+	"context"
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/autoscaler/cluster-autoscaler/metrics"
+	kube_client "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	podv1 "k8s.io/kubernetes/pkg/api/v1/pod"
+)
+
+const maxPodChangeAge = 10 * time.Second
+
+var (
+	podsResource             = "pods"
+	unschedulablePodSelector = fields.ParseSelectorOrDie("spec.nodeName==" + "" + ",status.phase!=" +
+		string(apiv1.PodSucceeded) + ",status.phase!=" + string(apiv1.PodFailed))
+)
+
+// StaticAutoscaler exposes recent autoscaler activity
+type StaticAutoscaler interface {
+	LastScaleUpTime() time.Time
+	LastScaleDownDeleteTime() time.Time
+}
+
+// LoopTrigger object implements criteria used to start new autoscaling iteration
+type LoopTrigger struct {
+	podObserver  *UnschedulablePodObserver
+	scanInterval time.Duration
+	autoscaler   StaticAutoscaler
+}
+
+// NewLoopTrigger creates a LoopTrigger object
+func NewLoopTrigger(podObserver *UnschedulablePodObserver, autoscaler StaticAutoscaler, scanInterval time.Duration) *LoopTrigger {
+	return &LoopTrigger{
+		podObserver:  podObserver,
+		scanInterval: scanInterval,
+		autoscaler:   autoscaler,
+	}
+}
+
+// Wait waits for the next autoscaling iteration
+func (t *LoopTrigger) Wait(lastRun time.Time) {
+	sleepStart := time.Now()
+	defer metrics.UpdateDurationFromStart("loopWait", sleepStart)
+
+	// To improve scale-up throughput, Cluster Autoscaler starts new iteration
+	// immediately if the previous one was productive.
+	if !t.autoscaler.LastScaleUpTime().Before(lastRun) ||
+		!t.autoscaler.LastScaleDownDeleteTime().Before(lastRun) {
+		select {
+		case <-t.podObserver.unschedulablePodChan:
```

Comment:
So you're clearing the signal about unschedulable pod appearing here, but not about scan interval passing. Why are you treating some triggers differently than others? This will lead to wasting a loop that isn't needed every now and then, which is not terrible, but definitely something we could avoid.

Reply:
We're clearing just this one channel because it persists between loops, while the time.After() is re-created with each iteration. I don't think it will have the effect you're suggesting, but I'll actually test that, for now publishing the other responses.

Reply:
Ah, you're right, we don't make a call to `After()` at all in this branch, nvm my comment then :)